Search save

partoa opened the pull request:
Solves  #7. There are some things worth noting.

The following fields are not available in any API endpoints:
1. UpdateDate
2. CategoryIDPath
3. SellerSellerAccountType

For 1, I have used the CreationDate. For 2 I have generated from the categories list provided in the summary. 3 is not available at all. Remains blank.

I have also noticed that the API Key in use is hardcoded in getCredentials.py. I suspect that part of the code needs to be refactored.

I have also made some significant changes to searchItem.py. You can find details in the commit.

partoa committed:
[WIP] Several changes for search and save feature
 1- Modified .gitignore to ignore .env file
 2- Added searchSave.py. Currently looping through products
 3- searchItem.py modified to function as a library
 4- Changed the sandbox URL from item_summary to item

partoa committed:
Data is being pulled, processed and saved as per issue #7
Othe things this commit does are as follows:
  1- Renamed getItemDetails in searchItem to getItemSummary
     This calls the item_summary/search endpoint
  2- Added getItem in searchItem. This gets actual item details
  3- Added credentials argument to getItemSummary. This caches the
     credentials for perfomance. Subsequent calls to getItemSummary
     will not make the API call. This should probably be moved to
     getCredentials.py. @TODO
  4- Added fieldgroups argument to getItemSummary.
  5- Added return_raw to getItemSummary. This will return JSON
     without processing it.

partoa commented:
The functionality is in searchSave.py. It takes an argument. Run it for more information. Creates an output file which is in the same folder as the .XML file. Output file is .output.xml. There is no notification for the file name. That could be added.

partoa committed:
Using environment variables for configuration. To enable this:
   1- Use of multiple endpoints in the code removed. Only one endpoint
      is avalable for every action. Sandbox or production is determined by
      the subdomain environment variable.
   2- Introduction of a .env file
   3- Script for loading variables into the commandline added

partoa committed:
Adding example.env.
Had been ommited in previous commit

partoa commented:
The scripts now use environment variables. To use them, please remember to call `source load_env.sh` before running any scripts.

partoa committed:
Fixed github workflow to set proper environment variables.
The correct secret values need to be configured as well

partoa committed:
Ensured test is running and using environment variables

partoa committed:
Spelling correction
SECTRET to SECRET

partoa commented:
Made numerous changes to make the API endpoint and credentials configurable. You should be able to pull now. Please note the the secret variable names for environment variables are now as follows:
APP_ID=SET-THIS
SECRET=SET-THIS
SUBDOMAIN=api.sandbox.ebay.com

ArtemGr commented:
TY. See if you can match the variables already defined in the action: https://github.com/mercatos/Ebay/blob/b79b8f53d6056a4f9324ff2e439fc9d5477609a5/.github/workflows/python-app.yml#L33

There's also an existing Python code for working with these.

ArtemGr commented on .github/workflows/python-app.yml:
```diff
@@ -30,7 +30,8 @@ jobs:
-        EBAY_SANDBOX: ${{ secrets.EBAY_SANDBOX }}
+        APP_ID: ${{ secrets.APP_ID }}
+        SECRET: ${{ secrets.SECRET }}
+        SUBDOMAIN: ${{ secrets.SUBDOMAIN }}
```
We have EBAY_SANDBOX_ID and EBAY_SANDBOX configured already (in GitHub and at work).
SECRET is less specific than EBAY_SANDBOX.
SUBDOMAIN can be derived from the presence of EBAY_PRO_ID.
Update the code to use the existing variable names please.

partoa replied:
Please share the specific values that I can use for each. SUBDOMAIN I can hardcode because it should always be the sandbox subdomain. APP_ID and SECRET I would need guidance on. Alternatively, I can return the code to what it was before. We may need to have a conversation about how best to handle this. There's a problem in that the code in the ebay-apis folder was working differently from what is in the root folder. In fact, I did not even need to touch the ebay-apis folder.

I will also add a unit test for the code that I created.

partoa replied:
Maybe you share a sample .env file. Or bash file with export commands. That way, I can have a replica environment for testing.

ArtemGr replied:
You add the ~~unit~~ integration test and I'll patch the env back.

ArtemGr reviewed (changes requested):
Add a test please that would demostrate the code working.
Include a sample of test.data.xml if necessary.

p.s. "*Secrets are not passed to workflows that are triggered by a pull request from a fork*", but I will likely run the GitHub Action test in a branch (https://github.com/mercatos/Ebay/tree/search_save).

partoa committed:
Added unit test for searchSave.py. Left everything else as is.
 - Using mockfs so as to not write anything into the file system
 - Using actuall output from the API and an expected output that takes into account
   the conditions.

partoa committed:
Better white space removal for unit test_searchSave
It's necessary to remove all white spaces to compare the strings. Alrternitive XML canonicalization.
This can be considered in future

partoa committed:
Added comment for future XML canonicalization consideration

partoa commented:
Added a unit test in tests/test_searchSave.py

The input XML, as per issue #7 is there. Also expected output, as per the API pickled API result data is there. You may add more data, for better correctness check.

ArtemGr commented:
I wanted to go over the code in order to reduce the future ownership costs, but with the tests the code is too large for me even to parse.

(For reference, the tests should demonstrate the code in a KISS manner, similar to http://sscce.org/).

Will keep it in the `search_save` branch.

Thanks for your effort, @partoa.

partoa commented:
The test itself is rather short. The data to test it is what is a lot. Could change it to JSON rather than serialized request objects. That might work better.

partoa commented:
Let me know if changing the test data for mocking will help things along.

ArtemGr commented:
> The test itself is rather short.

You've got me there: the test is too large in terms of the ratio between the noise and the code I'd wanna keep. Mocking HTTP response is completely unnecessary, because
a) In the integration test we want the code to reach the Ebay APIs - that is the very integration we're testing!
b) Tests should demonstrate how the code is used. Mocking turns the tests from being descriptive to being cryptic.

I'd recommend you "Test-Driven Development by Example" by Kent Beck if you're curious about the original ideas behind testing, now shadowed by a different tradition that spawned likely from the "code coverage" metrics.

*You show me anything that depicts institutional progress in America: school test scores, crime stats, arrest reports, anything that a politician can run on, anything that someone can get a promotion on. And as soon as you invent that statistical category, fifty people in that institution will be at work trying to figure out a way to make it look as if progress is occurring when actually none is. David Simon*

*The more any quantitative social indicator is used for social decision-making, the more subject it will be to corruption pressures and the more apt it will be to distort and corrupt the social processes it is intended to monitor. Campbell's law*

> Let me know if changing the test data for mocking will help things along.

No, thanks, I'm good. @fropl should release the funds.

partoa commented:
That is meant to be a Unit Test, not an integration test.

Nonetheless, should you ever come up with a standard for adding tests on this repo, and should I ever work on something else. I will follow your guidelines. For either. Beyond that, I'll let this one be. I had to come up with something being that there were no unit or integration tests in the repo to begin with. At least nothing that tests for correctness.

fropl commented:
@partoa @ArtemGr  Hello everyone and thank you! Made a payment!

partoa commented:
@fropl Super. @ArtemGr I'll be more than happy to engage further. Hopefully, at some point, there will be guidelines, and maybe some unit/integration tests that can be used as an additional guide. Guide by example, not just references to books.

ArtemGr commented:
> and maybe some unit/integration tests that can be used as an additional guide. Guide by example

There's a working test harness, https://github.com/mercatos/Ebay/runs/7917197850.
I was wondering why you would break it first and then reinvent the wheel.

Good point on guidelines. I did some notes to myself:

![image](https://user-images.githubusercontent.com/41847/188866706-bf4dcf6c-ab65-4ff3-9e9e-fc0799a9945c.png)

Shoot if you'd think of something to add.

partoa commented:
Noted on the guidelines. If you put them in the repo it would be fine. Word of advice, the first line sounds like an ad hominem. 

Discourse, in any team, and especially on Open Source projects is very important. You have to be able to have arguments. Set the rules. But no one will know what you want unless you set them and articulate them clearly. Otherwise, you'll dismiss all help you ever get and end up having to do everything yourself... In that case, why not just make a private project?

As for guidelines, please look at these for Metabase, I really like them. Clear, articulate and concise.
https://github.com/metabase/metabase/wiki/Writing-a-Driver%3A-Adding-Test-Extensions,-Tests,-and-Setting-up-CI
https://github.com/metabase/metabase/wiki/Metabase-Clojure-Style-Guide

Additionally, this is some unsolicited advice. Still on that first line.
There's another problem there, not sure what the name is. It's the kind that where there's a hierarchy, the higher-ups will end up doing everything that the team should do; because they don't trust anyone but themselves. Applies to everything too. Not just coding. If you are to lead, and you have the knowledge, you must learn how to lift your team up, not put them down.

Outside of that, I'm happy to contribute again if and when there is better discourse and more clarity on what is expected. Especially on code quality and proper framework around testing.

ArtemGr commented:
Regarding the Open Source, I've given it some thought ( https://observablehq.com/@artemgr/open-source-bounty-guide ), but bounty hunting is quite different from that. While we can employ some of the Open Source methodologies, not all of them are equally necessary or effective. (I'm still experimenting with these.) Thanks for the pointers though!

ArtemGr commented:
More notes. Thanks for the boost.

![image](https://user-images.githubusercontent.com/41847/188932519-ebedbf9f-f865-4210-8d42-10d32427cd0a.png)

ArtemGr commented:
> Discourse, in any team, and especially on Open Source projects is very important. You have to be able to have arguments.

Discourse is a fancy term, but irrelevant.
Per Ginnett, effective team can operate seamlessly across a range of authority (Laissez-faire - Democratic - Participative - Consultative - Autocratic).
Per Kahneman's “Focusing Illusion”, by promoting just the discussions it's easy to forget that the other modes are no less important.

Now, it's interesting to compare a predominantly democratic ecosystem (such as Rust) with Proof Of Work: It has to burn a lot of work in order to reach a quorum. (Meaning that it is neither eco-friendly nor sustainable.)

In Freelance and Bounty Hunting a shift towards speed is implied. We're adding some ~~money~~ [nitro](https://asphalt.fandom.com/wiki/Nitro) into the equation in order to speed things up (both on the sending **and on the receiving end**, making http://www.sscce.org/ particularly useful). Consequently, we expect to waste less work on discussion.

To borrow from https://dominiccummings.files.wordpress.com/2018/09/20180904-arpa-parc-paper1.pdf, in order to spend less time on discussions we could assume the “*tolerant interest and disinterest*” stance. But in the context of Freelance that means giving money away. Which Licklider did, in the millions.

If we don't have that kind of money just yet, we might prefer the “vertical integration”: woking in geographically proximate groups, - a path picked by Elon Musk, for example.

That is, the probability of a win is higher when communication is facilitated by the physical proximity, and per the [Kelly criterion](https://en.wikipedia.org/wiki/Kelly_criterion) the amount of time and money invested in a given interaction should be adjusted accordingly.

(p.s. The “*tolerant interest and disinterest*” stance reminds me also of [this fine quote](https://www.scottdmiller.com/wp-content/uploads/2020/03/Dallenback-1955-Phrenology-vs-psychoanalysis.pdf): "*Controversy over theory is futile. It advertises and calls attention to the theory criticized and it is not effective against the will to believe. Besides relieving critic's pent-up emotion, a polemical article probably accomplishes nothing; and the time, thought, and effort given to its writing are wasted. This is true because theories do not succumb to abstract argument. Criticisms are not read or, if read, are not assimilated or heeded by those wedded to the theory attacked-and they are not needed by those who are not ... I have searched long through the history of our science without finding a single instance in which I could with any assurance say that criticism was the coup de mort of theory. Theories pass from scientific stage not because they have been disproved but because they have been superseded-pushed off and replaced by others*".
While the discussion per se can be sourced for creativity, seeking a quorum is no less a waste than "*inventory is a waste*".)
